File: day 08/day8.cs
using System;
using System.Collections.Generic;
using System.IO;

namespace AdventProject
{
  class AdventOfCode
  {

    static long run(List<KeyValuePair<string, int>> instructions)
    {
      List<bool> visited = new List<bool>();
      for (int i = 0; i < instructions.Count; i++)
      {
        visited.Add(false);
      }
      long acc = 0;
      for (int i = 0; i != instructions.Count; i++)
      {
        if (visited[i])
        {
          return acc;
        }
        visited[i] = true;

        string instr = instructions[i].Key;
        int val = instructions[i].Value;

        if (instr == "nop")
        {
          continue;
        }
        else if (instr == "acc")
        {
          acc += val;
        }
        else if (instr == "jmp")
        {
          i += val - 1;
        }
      }
      Console.WriteLine("Part 2: " + acc);
      return acc;
    }
	
    static void Main(string[] args)
    {
      string inputFile = Directory.GetParent(Directory.GetCurrentDirectory()).Parent.Parent.FullName + @"\in.txt";
      StreamReader cin = new StreamReader(inputFile);

      string instr;
      int value;
      List<KeyValuePair<string, int>> instructions = new List<KeyValuePair<string, int>>();

      while (!cin.EndOfStream)
      {
        instr = cin.ReadLine();
        var index = instr.IndexOf(" ");
        value = int.Parse(instr.Substring(index, instr.Length - index));
        instr = instr.Substring(0, index);
        KeyValuePair<string, int> pair = new KeyValuePair<string, int>();
        instructions.Add(new KeyValuePair<string, int>(instr, value));
      }
	  
      Console.WriteLine("Part 1: " + run(instructions));

      var instrCopy = instructions;
      for (int i = 0; i < instructions.Count; i++)
      {
        if (instructions[i].Key == "jmp")
        {
          instrCopy[i] = new KeyValuePair<string, int>("nop", instructions[i].Value);
        }
        else if (instrCopy[i].Key == "nop")
        {
          instrCopy[i] = new KeyValuePair<string, int>("jmp", instructions[i].Value);
        }

        run(instrCopy);

        if (instructions[i].Key == "jmp")
        {
          instrCopy[i] = new KeyValuePair<string, int>("nop", instructions[i].Value);
        }
        else if (instrCopy[i].Key == "nop")
        {
          instrCopy[i] = new KeyValuePair<string, int>("jmp", instructions[i].Value);
        }
      }

    }
  }
}

File: day 08/day8.java
import java.io.File;
import java.io.IOException;
import java.util.*;

class Pair {
	public Pair(String instr, int value) {
		Instr = instr;
		Value = value;
	}

	public String Instr;
	public Integer Value;
};

public class AdventOfCode {

	static long run(Vector<Pair> v) {
		Vector<Integer> visited = new Vector<Integer>();/* (v.size(), 0); */
		for (int i = 0; i < v.size(); i++) {
			visited.add(0);
		}
		Long acc = (long) 0;
		for (int i = 0; i != v.size(); i++) {
			if (visited.get(i) != 0) {
				return acc;
			}
			visited.set((i), 1);

			String instr = v.get(i).Instr;
			int val = v.get(i).Value;

			if (instr == "nop") {
				continue;
			} else if (instr.equals("acc")) {
				acc += val;
			} else if (instr.equals("jmp")) {
				i += val - 1;
			}
		}
		System.out.println("Part 2: " + acc);
		return acc;
	}

	public static void main(String[] args) throws IOException {

		File file = new File("D:\\Java\\AdventOfCode\\in.txt");
		Scanner cin = new Scanner(file);

		String instr;
		int value;
		Vector<Pair> instructions = new Vector<Pair>();
		while (cin.hasNext()) {

			instr = cin.next();

			value = cin.nextInt();

			instructions.add(new Pair(instr, value));
		}
		System.out.println("Part 1: " + run(instructions));

		for (Pair i : instructions) {
			if (i.Instr.equals("jmp")) {
				i.Instr = "nop";
			} else if (i.Instr.equals("nop")) {
				i.Instr = "jmp";
			}

			run(instructions);

			if (i.Instr.equals("jmp")) {
				i.Instr = "nop";
			} else if (i.Instr.equals("nop")) {
				i.Instr = "jmp";
			}
		}

	}
}
